### PR TITLE
Replace script with build script, use features

### PR DESCRIPTION
**Pull request by ice1000**
See https://github.com/pingcap/kvproto/pull/349

Part of https://github.com/pingcap/raft-rs/issues/177

Signed-off-by: ice1000 <ice1000kotlin@foxmail.com>

### CODE REVIEW

**Comment by ice1000** [comment]
@BusyJay @overvenus PTAL

**Comment by ice1000** [comment]
Diff in /home/travis/build/pingcap/raft-rs/build.rs at line 137:
```diff
         let text = regex.replace_all(
             &text,
             "if $1 == ::protobuf::wire_format::WireTypeVarint {\
-                $3 = $2.read_enum()?;\
+             $3 = $2.read_enum()?;\
              } else {\
-                return ::std::result::Result::Err(::protobuf::rt::unexpected_wire_type(wire_type));\
+             return ::std::result::Result::Err(::protobuf::rt::unexpected_wire_type(wire_type));\
              }",
         );
```

Why? It's asking me to generate some codes that will not pass the fmt check.
Ridiculous

**Comment by nrc** [comment]
> Why? It's asking me to generate some codes that will not pass the fmt check.
Ridiculous

Bug in Rustfmt :-(

I think it is best to use `#[rustfmt::skip]` on this code

**Comment by ice1000** [comment]
> OTOH if this is meant to just be temporary, then I think it is OK.

Does creating a separated library for sharing the codes in this build script sounds like a good idea?

**Comment by nrc** [comment]
> Does creating a separated library for sharing the codes in this build script sounds like a good idea?

It does to me

**Comment by ice1000** [comment]
> It does to me

Let's listen to @siddontang's idea

**Comment by siddontang** [comment]
before we finish all things, don't introduce prost in the project. 

We need to verify in our own forked repo.

**Comment by ice1000** [comment]
@siddontang I've removed all prost dependencies, please take a look again.

File: build.rs [hunk 1]
```diff
@@ -45,28 +38,36 @@ fn main() {
                 e.expect("Couldn't list file").file_name().to_string_lossy()
             )
         })
+        .map(|s| &**s)
         .collect();
-    let file_names: Vec<_> = file_names.iter().map(|s| &**s).collect();
```
Comment (nrc) [inline build.rs]: :+1: good catch

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            (Some(_), Some(_)) | (None, None) => {
+                panic!("You must use exactly one of `lib-rust-protobuf` and `prost-buf` features")
+            }
+            (Some(_), _) => BufferLib::Prost,
+            (_, Some(_)) => BufferLib::Protobuf,
+        }
+    }
+}
+
+fn check_protoc_version() {
```
Comment (Hoverbear) [inline build.rs]: It's kind of unsatisfying to need to add a bash script for this purpose. This wholesale breaks our compatibility with systems without Bash like Windows. Do you think we can use `Command::new` and get the version from the `build.rs`?

File: Cargo.toml [hunk 3]
```diff
@@ -10,14 +10,22 @@ homepage = "https://github.com/pingcap/raft-rs"
...
+[build-dependencies]
+protoc-rust = "2.2.4"
+protoc-grpcio = "0.3.1"
+regex = "1.1"
 
 [features]
-default = []
+default = ["lib-rust-protobuf"]
+lib-rust-protobuf = ["protobuf"]
+regenerate = []
```
Comment (Hoverbear) [inline Cargo.toml]: This needs some documentation in `lib.rs` please. :)

Please make sure the user knows how to accomplish the main use cases this provides. Eg regenerating the protobufs.

Comment (ice1000) [reply Cargo.toml]: Actually I'm having trouble on using features, the `BufferLib::from_env_vars()` in build.rs does not seem to return `BufferLib::Prost` when I build this with `cargo build --features lib-prost`.

Comment (ice1000) [reply Cargo.toml]: FYI the `lib-prost = [...]` is removed because we don't introduce any prost things in this PR.

File: build.rs (hunk 2 above)
Comment (ice1000) [reply build.rs]: It's just some regex matching, maybe completely rewrite with Rust.

File: Cargo.toml (hunk 3 above)
Comment (ice1000) [reply Cargo.toml]: > This needs some documentation in `lib.rs` please. :)

How about `README.md` 😉?

Comment (Hoverbear) [reply Cargo.toml]: Both might be appropriate, but README.md is not part of generated docs.rs documentation. :(

**Comment by siddontang** [comment]
@ice1000 

I think you can create a `prost` branch later, then we can introduce prost.

**Comment by ice1000** [comment]
> @ice1000
> 
> I think you can create a `prost` branch later, then we can introduce prost.

This is what I'm doing right now. I have a local dev branch dealing with prost and I'm trying my best to reduce prost-related things in this PR.

File: build.rs [hunk 4]
```diff
@@ -30,13 +30,6 @@ fn main() {
         return;
     }
 
-    let buf_lib = BufferLib::from_env_vars();
-    if buf_lib == BufferLib::Prost {
-        unimplemented!("Prost support is not yet implemented");
-    }
-
-    check_protoc_version();
```
[the remark this answers is not in the document]
Comment (ice1000) [reply build.rs]: What do you mean by "both"? Protobuf/Prost or Generate/Non-generate?

Prost does not depend on protoc and I don't think we need protoc if we don't re-generate protos.

File: build.rs (hunk 2 above)
Comment (Hoverbear) [reply build.rs]: Please do. I prefer it if this was part of the PR.

File: build.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hoverbear) [reply build.rs]: Depending on `protoc` is highly undesirable.

File: Cargo.toml (hunk 3 above)
Comment (ice1000) [reply Cargo.toml]: I'll update README in prost PR. I don't think it's proper to introduce not-completely-ready features in README.

File: build.rs (hunk 2 above)
Comment (ice1000) [reply build.rs]: Give me some time to practice rust regex.

**Comment by ice1000** [comment]
Addressed @Hoverbear@'s and @overvenus@'s comments.

One task left: rewrite common.sh with rust

File: build.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ice1000) [reply build.rs]: Yes and I don't think we are.
Resolved

File: Cargo.toml (hunk 3 above)
Comment (Hoverbear) [reply Cargo.toml]: OK.

**Comment by Hoverbear** [comment]
@ice1000 Thanks for addressing this. It can make life easier for our downstream users.

**Comment by ice1000** [comment]
> @ice1000 Thanks for addressing this. It can make life easier for our downstream users.

Yes and this should be push to kvproto as well.

**Comment by ice1000** [comment]
Address all comments, @Hoverbear @overvenus @nrc PTAL

**Comment by ice1000** [comment]
I don't understand. Even if I apply this patch:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index ee64c47..0edd40e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -20,6 +20,7 @@ regex = "1.1"
 [features]
 default = ["lib-rust-protobuf"]
 lib-rust-protobuf = ["protobuf"]
+lib-prost = []
 regenerate = []
 failpoint = ["fail"]

diff --git a/build.rs b/build.rs
index 553a1cf..d9b8045 100644
--- a/build.rs
+++ b/build.rs
@@ -47,7 +47,7 @@ fn main() {

     match BufferLib::from_env_vars() {
         BufferLib::Prost => {
-            unimplemented!("Prost support is not yet implemented");
+            panic!("Prost support is not yet implemented");
         }

         BufferLib::Protobuf => {
@@ -84,7 +84,7 @@ impl BufferLib {
             env::var_os("CARGO_FEATURE_LIB_RUST_PROTOBUF"),
         ) {
             (Some(_), Some(_)) | (None, None) => {
-                panic!("You must use exactly one of `lib-rust-protobuf` and `prost-buf` features")
+                panic!("You must use exactly one of `lib-rust-protobuf` and `lib-prost` features")
             }
             (Some(_), _) => BufferLib::Prost,
             (_, Some(_)) => BufferLib::Protobuf,
```

and run `cargo build --features lib-prost`, it builds successfully. Does that mean my "panic" is not working?

**Comment by overvenus** [comment]
Try `cargo build --features "lib-prost,regenerate"`.

BTW, You might be interested in the cargo books, it explains rust build script in great detail.  https://doc.rust-lang.org/cargo/reference/build-scripts.html

**Comment by Hoverbear** [comment]
@nrc Regarding Raft <-> KVProto interaction, I think having Raft depend on KVProto isn't a terrible idea, but this would mean (like for the client) we'd need to publish it. https://github.com/pingcap/kvproto/issues/280

**Comment by ice1000** [comment]
```
error: failed to select a version for `protobuf`.
    ... required by package `raft v0.5.0 (/home/travis/build/pingcap/raft-rs)`
versions that meet the requirements `~2.0-2.2` are: 2.0.6, 2.0.5, 2.0.4, 2.0.3, 2.0.2, 2.0.1, 2.0.0
all possible versions conflict with previously selected packages.

  previously selected package `protobuf v2.2.4`
    ... which is depended on by `protoc-rust v2.2.4`
    ... which is depended on by `raft v0.5.0 (/home/travis/build/pingcap/raft-rs)`
```

Why..

**Comment by Hoverbear** [comment]
@ice1000 I can only offer limited help without a reproduction case. :(

It seems that because on Master we went from Protobuf 2.2.4 to Protobuf ~2.0-2.2 it broke for you. For release versions it should have gone from 2.0 to ~2.0-2.2 so it wouldn't break.

But this works:

```
[hoverbear@Obsidian raft-rs]$ cargo clean
[hoverbear@Obsidian raft-rs]$ cargo update
    Updating crates.io index
    Updating aho-corasick v0.6.9 -> v0.6.10
    Updating serde v1.0.87 -> v1.0.88
    Updating serde_derive v1.0.87 -> v1.0.88
[hoverbear@Obsidian raft-rs]$ cargo build
   Compiling unicode-xid v0.1.0
   Compiling libc v0.2.48
   Compiling byteorder v1.3.1
   Compiling rand_core v0.4.0
   Compiling protobuf v2.0.6
   Compiling cfg-if v0.1.6
   Compiling scopeguard v0.3.3
   Compiling quick-error v1.2.2
   Compiling log v0.4.6
   Compiling proc-macro2 v0.3.8
   Compiling rand_core v0.3.1
   Compiling quote v0.5.2
   Compiling hashbrown v0.1.8
   Compiling rand v0.5.6
   Compiling syn v0.13.11
   Compiling getset v0.0.6
   Compiling raft v0.5.0 (/mnt/c/Users/Hoverbear/Git/raft-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 13.02s
[hoverbear@Obsidian raft-rs]$
```

**Comment by BusyJay** [comment]
I don't think it's appropriate to make raft-rs depend on kvproto. raft-rs is supposed to be generic and minimize dependency.

**Comment by ice1000** [comment]
At least I'm fine with both case. The only breaking change in this PR is the location of `eraftpb` is moved to `rsprotobuf::eraftpb`.

**Comment by ice1000** [comment]
At least I'm fine with both case. The only breaking change in this PR is the location of `eraftpb` is moved to `rsprotobuf::eraftpb`.

**Comment by ice1000** [comment]
So the current status of this PR is "blocked by protobuf dependency version selection"?

**Comment by ice1000** [comment]
Resolved merge conflicts.

**Comment by Hoverbear** [comment]
@ice1000 There are new conflicts from #184 . :( There is a merge resolve here: https://github.com/ice1000/raft-rs/pull/1 you can accept.

**Comment by ice1000** [comment]
Can we have a merge please?

**Comment by Hoverbear** [comment]
@nrc is your LGTM still valid? :)

**Comment by nrc** [comment]
@ice1000 could you sort out the commit series please? Would be great to be based on master and not have the merge commits as well as having a coherent set of commits.

Could you use [protobuf-build](https://github.com/nrc/kvproto-build)? Then we wouldn't need to depend on kvproto or duplicate code.

**Comment by ice1000** [comment]
> @ice1000 could you sort out the commit series please?

Yeah sure

**Comment by ice1000** [comment]
`protobuf-build` is cool, will do.

**Comment by Hoverbear** [comment]
@nrc Is there a plan to publish that? Raft is a published crate, so it's dependencies must be as well.

**Comment by nrc** [comment]
> @nrc Is there a plan to publish that? Raft is a published crate, so it's dependencies must be as well.

Yes, I think once we're ready to merge all the Prost work we can move it to PingCAP org and publish. For now it depends on a modified Prost so we can't.

**Comment by Hoverbear** [comment]
@nrc So you'd like to depend on it and prevent us from publishing a new release until the prost situation is solved?

**Comment by nrc** [comment]
> @nrc So you'd like to depend on it and prevent us from publishing a new release until the prost situation is solved?

Yeah, I'm hopeful that won't take too long. We can make it work without the Prost dep if we need to publish a release before Prost does.

**Comment by Hoverbear** [comment]
Do you think we can just keep work to this branch while we get there?

**Comment by ice1000** [comment]
Yes.

**Comment by nrc** [comment]
Closing in favour of #201

**Pull request closed without merging**